### ISSUE TITLE
Watch requests for specific resources fail to return data (sometimes)

### ISSUE THREAD

**Issue by evanphx** [kubernetes/kubernetes#16699]
* Given an service named `jpl-input`
* A request (call it request A) is issued: `curl localhost:8001/api/v1/watch/namespaces/default/endpoints/jpl-input`.
* A request (call it request B) is issued: `curl localhost:8001/api/v1/watch/namespaces/default/endpoints`
* A pod that matches the service `jpl-input` is then started (either directly or via a rolling-update).
* Request B begins getting data: `{"type":"MODIFIED","object":{"kind":"Endpoints","apiVersion":"v1","metadata":{"name":"jpl-input","namespace":"default","sel...`
* Request A stays silent and never returns any data.

The expectation is that request A is sent updates about the endpoint.

Here is the tcp session between apiserver and etcd for the processing of request A:

```
root@ip-172-20-0-9:~# cat watch-on-specific-name 
input: watch.dump
filter: (ip or ip6) and ( port 52336 )
match: .
#######
T 127.0.0.1:52336 -> 127.0.0.1:4001 [AP]
  GET /v2/keys/registry/services/endpoints/default/jpl-input?quorum=false&recursive=false&sorted=false HTTP/1.1..Host: 127.0.0.1:4
  001..User-Agent: Go 1.1 package http..Accept-Encoding: gzip....                                                                 
#
T 127.0.0.1:4001 -> 127.0.0.1:52336 [AP]
  HTTP/1.1 200 OK..Content-Type: application/json..X-Etcd-Cluster-Id: 7e27652122e8b2ae..X-Etcd-Index: 1809352..X-Raft-Index: 13467
  978..X-Raft-Term: 5..Date: Tue, 03 Nov 2015 00:16:35 GMT..Content-Length: 808....{"action":"get","node":{"key":"/registry/servic
  es/endpoints/default/jpl-input","value":"{\"kind\":\"Endpoints\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"jpl-input\",\"na
  mespace\":\"default\",\"selfLink\":\"/api/v1/namespaces/default/endpoints/jpl-input\",\"uid\":\"c2b1ff23-6eb8-11e5-945c-024ba4ed
  8969\",\"resourceVersion\":\"1809296\",\"creationTimestamp\":\"2015-10-09T19:05:52Z\",\"labels\":{\"name\":\"jpl-input\"}},\"sub
  sets\":[{\"addresses\":[{\"ip\":\"10.244.0.65\",\"targetRef\":{\"kind\":\"Pod\",\"namespace\":\"default\",\"name\":\"formatter-0
  50683415242a28fb581d52fe0e42834-7zcpo\",\"uid\":\"cd2e2900-81bf-11e5-9bf3-024ba4ed8969\",\"resourceVersion\":\"1809295\"}}],\"po
  rts\":[{\"name\":\"jpl-input\",\"port\":9101,\"protocol\":\"TCP\"}]}]}","modifiedIndex":1809318,"createdIndex":1172435}}.       
##
T 127.0.0.1:52336 -> 127.0.0.1:4001 [AP]
  GET /v2/keys/registry/services/endpoints/default/jpl-input?wait=true&waitIndex=1809353 HTTP/1.1..Host: 127.0.0.1:4001..User-Agen
  t: Go 1.1 package http..Accept-Encoding: gzip....                                                                               
#
T 127.0.0.1:4001 -> 127.0.0.1:52336 [AP]
  HTTP/1.1 200 OK..Content-Type: application/json..X-Etcd-Cluster-Id: 7e27652122e8b2ae..X-Etcd-Index: 1809352..X-Raft-Index: 13467
  978..X-Raft-Term: 5..Date: Tue, 03 Nov 2015 00:16:35 GMT..Transfer-Encoding: chunked....                                        
########################################
T 127.0.0.1:4001 -> 127.0.0.1:52336 [AP]
  731..{"action":"compareAndSwap","node":{"key":"/registry/services/endpoints/default/jpl-input","value":"{\"kind\":\"Endpoints\",
  \"apiVersion\":\"v1\",\"metadata\":{\"name\":\"jpl-input\",\"namespace\":\"default\",\"selfLink\":\"/api/v1/namespaces/default/e
  ndpoints/jpl-input\",\"uid\":\"c2b1ff23-6eb8-11e5-945c-024ba4ed8969\",\"resourceVersion\":\"1809318\",\"creationTimestamp\":\"20
  15-10-09T19:05:52Z\",\"labels\":{\"name\":\"jpl-input\"}},\"subsets\":[{\"addresses\":[{\"ip\":\"10.244.0.65\",\"targetRef\":{\"
  kind\":\"Pod\",\"namespace\":\"default\",\"name\":\"formatter-050683415242a28fb581d52fe0e42834-7zcpo\",\"uid\":\"cd2e2900-81bf-1
  1e5-9bf3-024ba4ed8969\",\"resourceVersion\":\"1809295\"}},{\"ip\":\"10.244.2.85\",\"targetRef\":{\"kind\":\"Pod\",\"namespace\":
  \"default\",\"name\":\"formatter-050683415242a28fb581d52fe0e42834-ucva4\",\"uid\":\"29dc945a-81c0-11e5-9bf3-024ba4ed8969\",\"res
  ourceVersion\":\"1809378\"}}],\"ports\":[{\"name\":\"jpl-input\",\"port\":9101,\"protocol\":\"TCP\"}]}]}","modifiedIndex":180937
  9,"createdIndex":1172435},"prevNode":{"key":"/registry/services/endpoints/default/jpl-input","value":"{\"kind\":\"Endpoints\",\"
  apiVersion\":\"v1\",\"metadata\":{\"name\":\"jpl-input\",\"namespace\":\"default\",\"selfLink\":\"/api/v1/namespaces/default/end
  points/jpl-input\",\"uid\":\"c2b1ff23-6eb8-11e5-945c-024ba4ed8969\",\"resourceVersion\":\"1809296\",\"creationTimestamp\":\"2015
  -10-09T19:05:52Z\",\"labels\":{\"name\":\"jpl-input\"}},\"subsets\":[{\"addresses\":[{\"ip\":\"10.244.0.65\",\"targetRef\":{\"ki
  nd\":\"Pod\",\"namespace\":\"default\",\"name\":\"formatter-050683415242a28fb581d52fe0e42834-7zcpo\",\"uid\":\"cd2e2900-81bf-11e
  5-9bf3-024ba4ed8969\",\"resourceVersion\":\"1809295\"}}],\"ports\":[{\"name\":\"jpl-input\",\"port\":9101,\"protocol\":\"TCP\"}]
  }]}","modifiedIndex":1809318,"createdIndex":1172435}}...0....                                                                   
##
T 127.0.0.1:52336 -> 127.0.0.1:4001 [AP]
  GET /v2/keys/registry/services/endpoints/default/jpl-input?wait=true&waitIndex=1809380 HTTP/1.1..Host: 127.0.0.1:4001..User-Agen
  t: Go 1.1 package http..Accept-Encoding: gzip....                                                                               
#
T 127.0.0.1:4001 -> 127.0.0.1:52336 [AP]
  HTTP/1.1 200 OK..Content-Type: application/json..X-Etcd-Cluster-Id: 7e27652122e8b2ae..X-Etcd-Index: 1809379..X-Raft-Index: 13468
  044..X-Raft-Term: 5..Date: Tue, 03 Nov 2015 00:16:54 GMT..Transfer-Encoding: chunked....                                        
####################################################################################################################
T 127.0.0.1:4001 -> 127.0.0.1:52336 [AP]
  731..{"action":"compareAndSwap","node":{"key":"/registry/services/endpoints/default/jpl-input","value":"{\"kind\":\"Endpoints\",
  \"apiVersion\":\"v1\",\"metadata\":{\"name\":\"jpl-input\",\"namespace\":\"default\",\"selfLink\":\"/api/v1/namespaces/default/e
  ndpoints/jpl-input\",\"uid\":\"c2b1ff23-6eb8-11e5-945c-024ba4ed8969\",\"resourceVersion\":\"1809379\",\"creationTimestamp\":\"20
  15-10-09T19:05:52Z\",\"labels\":{\"name\":\"jpl-input\"}},\"subsets\":[{\"addresses\":[{\"ip\":\"10.244.2.85\",\"targetRef\":{\"
  kind\":\"Pod\",\"namespace\":\"default\",\"name\":\"formatter-050683415242a28fb581d52fe0e42834-ucva4\",\"uid\":\"29dc945a-81c0-1
  1e5-9bf3-024ba4ed8969\",\"resourceVersion\":\"1809378\"}}],\"ports\":[{\"name\":\"jpl-input\",\"port\":9101,\"protocol\":\"TCP\"
  }]}]}","modifiedIndex":1809401,"createdIndex":1172435},"prevNode":{"key":"/registry/services/endpoints/default/jpl-input","value
  ":"{\"kind\":\"Endpoints\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"jpl-input\",\"namespace\":\"default\",\"selfLink\":\"/
  api/v1/namespaces/default/endpoints/jpl-input\",\"uid\":\"c2b1ff23-6eb8-11e5-945c-024ba4ed8969\",\"resourceVersion\":\"1809318\"
  ,\"creationTimestamp\":\"2015-10-09T19:05:52Z\",\"labels\":{\"name\":\"jpl-input\"}},\"subsets\":[{\"addresses\":[{\"ip\":\"10.2
  44.0.65\",\"targetRef\":{\"kind\":\"Pod\",\"namespace\":\"default\",\"name\":\"formatter-050683415242a28fb581d52fe0e42834-7zcpo\
  ",\"uid\":\"cd2e2900-81bf-11e5-9bf3-024ba4ed8969\",\"resourceVersion\":\"1809295\"}},{\"ip\":\"10.244.2.85\",\"targetRef\":{\"ki
  nd\":\"Pod\",\"namespace\":\"default\",\"name\":\"formatter-050683415242a28fb581d52fe0e42834-ucva4\",\"uid\":\"29dc945a-81c0-11e
  5-9bf3-024ba4ed8969\",\"resourceVersion\":\"1809378\"}}],\"ports\":[{\"name\":\"jpl-input\",\"port\":9101,\"protocol\":\"TCP\"}]
  }]}","modifiedIndex":1809379,"createdIndex":1172435}}...0....                                                                   
##
T 127.0.0.1:52336 -> 127.0.0.1:4001 [AP]
  GET /v2/keys/registry/services/endpoints/default/jpl-input?wait=true&waitIndex=1809402 HTTP/1.1..Host: 127.0.0.1:4001..User-Agen
  t: Go 1.1 package http..Accept-Encoding: gzip....                                                                               
#
T 127.0.0.1:4001 -> 127.0.0.1:52336 [AP]
  HTTP/1.1 200 OK..Content-Type: application/json..X-Etcd-Cluster-Id: 7e27652122e8b2ae..X-Etcd-Index: 1809401..X-Raft-Index: 13468
  179..X-Raft-Term: 5..Date: Tue, 03 Nov 2015 00:17:50 GMT..Transfer-Encoding: chunked....                                        
###################exit
```

**Comment by caesarxuchao** [comment]
cc @lavalamp

**Comment by lavalamp** [comment]
Recommend switching to `localhost:8001/api/v1/namespaces/default/endpoints?watch=true&metadata.name=jpl-input`, although I'm not sure off the top of my head why what you're doing isn't working.

**Comment by evanphx** [comment]
@lavalamp that outputs updates for other endpoints as well: `{"type":"MODIFIED","object":{"kind":"Endpoints","apiVersion":"v1","metadata":{"name":"rabbitmq-api","namespace":"default" ...`

**Comment by evanphx** [comment]
I'm trying `curl 'localhost:8001/api/v1/namespaces/default/endpoints?watch=true&fieldSelector=metadata.name=jpl-input'` but that doesn't seem to be working either, same as request A.

**Comment by evanphx** [comment]
Yup, finished, using a fieldSelector results in no output, just like request A.

**Comment by lavalamp** [comment]
@caesarxuchao, can you double check that the endpoints registry is set up correctly? We should add a test.

**Comment by lavalamp** [comment]
Oh, well, that's interesting. (sorry I forgot the fieldSelector bit.)

Is our namespace handling broken? `curl 'localhost:8001/api/v1/namespaces/default/endpoints?watch=true&fieldSelector=metadata.name=jpl-input,metadata.namespace=default'` shouldn't make a difference, but...

**Comment by evanphx** [comment]
Well: `curl 'localhost:8001/api/v1/endpoints?watch=true&fieldSelector=metadata.name=jpl-input,metadata.namespace=default'`

Gives me: `field label not supported: metadata.namespace`

**Comment by lavalamp** [comment]
OK, that's what it's supposed to do. Hm.

**Comment by caesarxuchao** [comment]
@evanphx I tried again tonight on my local cluster and can't reproduce the error anymore. Perhaps it has something to do with the workload? Is your cluster running heavy workload?

**Comment by evanphx** [comment]
The cluster is in a steady state mostly, no cluster changes. The pods are doing stuff but nothing that I'd think would churn the API much (if anything kube-proxy is busy forwarding traffic and thats it).

**Comment by evanphx** [comment]
I looked at the traffic being sent to etcd again and while the system is steady, every 10 seconds or so, I see a request like `PUT /v2/keys/registry/minions/ip-172-20-0-215.us-west-2.compute.internal?prevIndex=1809350&prevValue=<encoded json>`. Looking at the responses to the PUTs, they all succeed so the horizon for etcd's versions is going forward pretty constantly.

**Comment by evanphx** [comment]
So, this turned out to be an issue with `kubectl proxy` not sending the data back! A quick `git bisect` session shows it was fixed in #11759. Thanks to @caesarxuchao and @lavalamp for humoring me and @vishh for fixing this bug back in July!
